### ISSUE TITLE
Coupling of authorisation service with session / launch context

### ISSUE THREAD

**Issue by whbath** [smart-on-fhir/smart-on-fhir.github.io#18]
Hi there,
Just questioning if there is an alternative to the options around launch scopes / what the thinking is around coupling the authorisation service with launch context.

The approach for passing launch context (via authorisation server as part of issuing an access token) seems based on the premise that our system can modify the behaviour of the authorisation server. We are utilising a 3rd party API Manager (and I assume a lot of others would look to do so as well). Conflating passing of session information / launch context in with this token means we need to look to modify a 3rd party product which we'd rather not do. Are there other options on the table to achieve similar without the underlying assumption that we can modify the authorisation server? E.g. a FHIR Other resource or similar where the lookup of launch scope can be done on the backend? I can imagine a lot of companies would look to use COTS API Managers and the current approach makes adoption of this hard for those of us that do.

Thanks,
Will.

**Comment by jmandel** [comment]
Hi Will,

Thanks for the question -- it's an important one for us to address. You're correct that we do couple our context-communication mechanism to the authorization process. And ideally it would be great if we didn't' need to. We'd be happy to consider alternatives, but let me provide some of the rationale for why we do things this way right now:

We want to allow apps to launch with context, and with user approval (authorization). Not all apps will require an explicit end-user approval process (that will depend on local policies, and the story may be different for clinician-facing vs. user-facing apps). But some will -- e.g. a patient-facing app that's asking a portal user to give it access to a patient record. In this scenario, let's say a mother Alice with two kids, Bob and Charlie, wants to authorizing a growth chart app to access Bob's record. So the launch context is effectively limiting the app to just access demographics and observations for one patient (Bob).

At the time of the authorization process, what is Alice clicking "okay" to? If the launch context (which binds the "current patient" to Bob's record) isn't already established, then Alice can't really give her okay, because she's not sure quite what she's consenting to. If the launch context has been established, then the Authorization server can very explicitly show Alice exactly what permissions she'd be approving (e.g. something like "Ahoy, Alice! You are about to share Bob's demographics + observations with the Growth Charts App. Do you want to proceed?).

So that's the rationale for coupling them.

But let's think outside of the box here. Are there ways to de-couple that still give Alice sufficient information on which to base an authorization decision?

**Comment by grahamegrieve** [comment]
I think that there are, but I very much doubt that using a COTS authorization server is a sustainable approach for serious use. But maybe for the development process it would help a lot?

**Comment by whbath** [comment]
Sorry for the radio silence this is something I've been thinking about and discussing.

First up, great context I now understand what you want to achieve with this which is great. In response to your question after consideration I can't think of a more elegant way to achieve those objectives; if you want to authorise at that level then you'd need your authorisation service to be aware of context up front. That said I think the intention is good but with all things there are tradeoffs and I'd be interested to thoughts around the drawbacks of attempting to do so.

Authorisation Fatigue: Would a person want to get that specific with authorisation; i.e. if I trust it to access the growth information for one child, why not the other? Have we considered the concept of "authorisation fatigue"? Often most users click through permission boilerplate, if you have trusted the application to access information for one child, do I really want to be re-prompted for every subsequent child? More so, if I have got a collection of SMART components/widgets on a page (say a half dozen), every time I open a new patient I will get half a dozen re-prompts for each patient? Most clinical users I've dealt with abhor an extra mouse click; let alone another dozen which will likely not be read and just clicked-through?

Information Privacy leaks / complexity in understand the domain: For the example case where launch context might reference specific Patient and Medications resources, for this to work the system has to be intelligent enough to pull back the data to construct authorisation prompts like "The drug ordering  application wants to access the prescription for <insert anti-psychotic drug> for <insert celebrities name here>". Hence your authorisation service then needs to be capable of interpreting quite a lot of clinical context to generate meaningful authorisation prompts? Additionally that in itself is telling the user a lot, to prevent clever users leaking information through prompts your authorisation service would then need to know your specific authorisation logic to apply it apriori to figure out if it should disclose the association of the celebrity with that drug in the authorisation prompts. Whereas for a general "Application X wants to access patient and drug information on your behalf" you're not disclosing anything?

Appreciate your thoughts / what I might be missing?

**Comment by grahamegrieve** [comment]
firstly, implementing an oauth service is actually simple. It's the authentication that's hard and what would be appropriate to off-load to COTS. There's a discussion of this here : http://www.healthintersections.com.au/?p=2108 - e.g. authenticate the user using your COTS solution, but then ask them the API access questions yourself, based on what you can do in the API. 

I would think that authorisation fatigue is dealt with by providing appropriate defaults

As for the question of the scopes and clinical context - yes, this is a very big question that needs considerable research before we can define interoperable scopes that are useful to patient / provider

**Comment by explunit** [comment]
> Not all apps will require an explicit end-user approval process (that will depend on local policies, and the story may be different for clinician-facing vs. user-facing apps).

Given the above, and since adding complexity to the Authorize endpoint necessarily increases its attack surface, shouldn't processing of the `launch` parameter be optional for EHRs? This would meant that SMART apps should be designed so they still function if all they have is initially-received launch information and interaction with standard OAuth2 and FHIR. 

EHRs will already have a well-defined set of rules around which users may see which data, and there is no need to add to the "alert fatigue".

Yes, this means more complexity in the code of the SMART apps. But from an industry-wide security perspective it seems more prudent than requiring all EHRs have a customized OAuth2 Authorize endpoint.

@whbath also raises some very good points about unintentional privacy leaks during the authentication/prompt process. To address them requires even more complexity (and thus a larger attack surface) in the initial authentication flow.

**Comment by jmandel** [comment]
It's a fair point that some EHRs may not support scopes like
"launch/patient" - and these EHRs will still be able to run SMART apps that
don't depend on the EHR handling patient selection.

But if we want to talk about attack surfaces, I think an important best
practice is to limit the amount of data we share with apps; and allowing
apps to request single-patient access is an important part of how we
accomplish this. In other words, I'd rather trust 100 EHRs to correctly
implement an authorization service than trust 10,000 apps to acquire
user-level, multi-patient access (if they don't strictly need it). It's not
just a question of "more complexity" for the apps; it's a question of more
*trust*.

That said, of course different features will make sense in different
environments. I'm happy to see experimentation here from real-world
systems!

**Comment by whitehatguy** [comment]
I'm fairly neutral on the topic of SHOULD versus MUST on EHR support for the client-initiated launch workflow, but I'd like to provide some feedback on your statement, @explunit: 

>Yes, this means more complexity in the code of the SMART apps. But from an industry-wide security perspective it seems more prudent than requiring all EHRs have a customized OAuth2 Authorize endpoint

SMART on FHIR is a profile of OAuth2 with it's own processing rules.  OAuth2 explicitly states the following:

>This framework was designed with the clear expectation that future work will define prescriptive profiles and extensions necessary to achieve full web-scale interoperability.

In this context, SMART on FHIR is a "prescriptive profile" that defines such extensions to achieve interoperability between healthcare applications and electronic health records.  While there may be off-the-shelf SDK's designed to help developers build OAuth2 profiles, I wouldn't have the expectation that an OAuth2 profile such as SMART on FHIR (or any other domain-specific profile) would be instantly compatible with a profile-agnostic "generic OAuth2" service.  There may be profiles in the wild that are being developed that could cover a wider range of verticals (Kantara Initiative's [UMA][1], for example) in the future.

I'm still thinking through the usability aspects for where user interactions must occur for either usability or security.  It feels like patient-facing workflows would be significantly different than provider-facing; at a minimum, I'm going to assume that the user's persona affects the mechanism in which they authenticate, the user experience of the authorization decision, and the business logic for when an application can proceed without explicit user consent.

[1]: https://docs.kantarainitiative.org/uma/draft-uma-core.html

**Comment by explunit** [comment]
@whitehatguy Does the concept of a Profile (in OAuth2 terms) mean a specific set of response_type, scope, property values, etc? Or can it also include completely new querystring parameters to OAuth2 standard endpoints, e.g. `launch`?

(Not a rhetorical question -- I couldn't find a clear answer of how far you can go with a profile and still call it OAuth)

**Comment by whitehatguy** [comment]
You're correct in that it wasn't explicitly defined as to what constitutes a "profile" in the RFC.  I can say that many of the "profiles" that have been released for OAuth2 tend to include new parameters or headers not defined by the original.  An example is the [JSON Web Token Profile for OAuth 2.0 Client Authentication and Authorization Grants][1].  In this profile, a client application sends an additional "assertion" query parameter whose value is a signed JWT as proof of authorization.  This parameter is formally declared in ["Assertion Framework for OAuth 2.0"][2].  There's also [OAuth 2.0 Message Authentication Code (MAC) Tokens][3] which attempts to define huge swaths of new protocol bits (it also appears to have expired.)

[1]: http://self-issued.info/docs/draft-ietf-oauth-jwt-bearer.html
[2]: https://tools.ietf.org/html/draft-ietf-oauth-assertions-05
[3]: https://tools.ietf.org/html/draft-ietf-oauth-v2-http-mac-05
